File: 0064-Minimum-Path-Sum/lc0064.cc
// C++ solution.

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        if (grid.empty()) { return 0; }
        int m = grid.size();
        int n = grid.front().size();

        // 1-d dp array is enough
        vector<int> dp(n, numeric_limits<int>::max());
        dp[0] = 0;
        for (int i = 0; i < m; i++) {
            dp[0] += grid[i][0];
            for (int j = 1; j < n; j++) {
                dp[j] = min(dp[j], dp[j-1]) + grid[i][j];
            }
        }
        return dp[n-1];
    }
};

File: 0064-Minimum-Path-Sum/lc0064.rs
// Rust solution.

impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        if grid.is_empty() { 0 }
        else {
            let (m, n) = (grid.len(), grid[0].len());
            let mut dp: Vec<i32> = vec![std::i32::MAX; n];
            dp[0] = 0;
            for i in 0..m {
                dp[0] += grid[i][0];
                for j in 1..n {
                    dp[j] = dp[j-1].min(dp[j]) + grid[i][j];
                }
            }
            dp[n-1]
        }
    }
}
